src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,83 @@
+
 public class Injector {
     private static final Injector injector = new Injector();
+    private final Map<Class<?>, Object> instances = new HashMap<>();
+    private final Map<Class<?>, Class<?>> implementationMap = new HashMap<>();
 
     public static Injector getInjector() {
         return injector;
     }
 
     public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+        if (!interfaceClazz.isAnnotationPresent(Component.class)) {
+            throw new RuntimeException(
+                    "Injection failed, missing @Component annotation on the class "
+                    + interfaceClazz.getName());
+        }
+        Object clazzImplementationInstance = null;
+        Class<?> clazz = findImplementation(interfaceClazz);
+        Field[] declaredFields = clazz.getDeclaredFields();
+
+        for (Field field: declaredFields) {
+            if (field.isAnnotationPresent(Inject.class)) {
+                Object fieldInstance = getInstance(field.getType());
+                clazzImplementationInstance = createNewInstance(clazz);
+                field.setAccessible(true);
+                try {
+                    field.set(clazzImplementationInstance, fieldInstance);
+                } catch (IllegalAccessException e) {
+                    throw new RuntimeException("Cant initialize field value. /n Class: "
+                            + clazz.getName() + "/n Field: " + field.getName() + "/n", e);
+                }
+            }
+        }
+
+        if (clazzImplementationInstance == null) {
+            clazzImplementationInstance = createNewInstance(clazz);
+        }
+
+        return clazzImplementationInstance;
+    }
+
+    private Object createNewInstance(Class<?> clazz) {
+        if (instances.containsKey(clazz)) {
+            return instances.get(clazz);
+        }
+        Constructor<?> constructor = null;
+        try {
+            constructor = clazz.getConstructor();
+            Object object = constructor.newInstance();
+            instances.put(clazz, object);
+            return object;
+        } catch (NoSuchMethodException
+                 | InstantiationException
+                 | IllegalAccessException
+                 | InvocationTargetException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private Class<?> findImplementation(Class<?> interfaceClazz) {
+        implementationMap.put(FileReaderService.class, FileReaderServiceImpl.class);
+        implementationMap.put(ProductParser.class, ProductParserImpl.class);
+        implementationMap.put(ProductService.class, ProductServiceImpl.class);
+
+        implementationMap.put(FileReaderServiceImpl.class, FileReaderServiceImpl.class);
+        implementationMap.put(ProductParserImpl.class, ProductParserImpl.class);
+        implementationMap.put(ProductServiceImpl.class, ProductServiceImpl.class);
+        return implementationMap.get(interfaceClazz);
```

Comment:
The `implementationMap` is being populated inside the `findImplementation` method. Consider moving this initialization to a static block or constructor to avoid repeated initialization every time this method is called.